File: validator/validator.go
package validator

import (
	"fmt"
	"os"
	"strings"

	"github.com/AndresCarvajalx/notiflow/config"
	"github.com/AndresCarvajalx/notiflow/excel"
	"github.com/xuri/excelize/v2"
)

type Result struct {
	Valid            bool
	FileExists       bool
	ColumnsMissing   []string
	TotalRows        int
	RowsWithPhone    int
	RowsWithoutPhone int
	SamplePhone      string
	Errors           []string
	Warnings         []string
}

func Run(cfg *config.Config) *Result {
	r := &Result{Valid: true}

	info, err := os.Stat(cfg.Excel.Path)
	if err != nil {
		r.Errors = append(r.Errors, fmt.Sprintf("Archivo no encontrado: %s", cfg.Excel.Path))
		r.Valid = false
		return r
	}
	if info.IsDir() {
		r.Errors = append(r.Errors, fmt.Sprintf("La ruta es un directorio: %s", cfg.Excel.Path))
		r.Valid = false
		return r
	}
	r.FileExists = true

	f, err := excelize.OpenFile(cfg.Excel.Path)
	if err != nil {
		r.Errors = append(r.Errors, fmt.Sprintf("Error abriendo Excel: %v", err))
		r.Valid = false
		return r
	}
	defer f.Close()

	sheet := f.GetSheetName(0)
	rows, err := f.GetRows(sheet)
	if err != nil {
		r.Errors = append(r.Errors, fmt.Sprintf("Error leyendo hoja '%s': %v", sheet, err))
		r.Valid = false
		return r
	}

	headerRow := cfg.Excel.HeaderRow
	if headerRow == 0 {
		headerRow = 1
	}
	headerIdx := headerRow - 1

	if len(rows) <= headerIdx {
		r.Errors = append(r.Errors, fmt.Sprintf("El archivo no tiene suficientes filas (tiene %d, necesita al menos %d)", len(rows), headerRow))
		r.Valid = false
		return r
	}

	actualHeaders := make(map[string]int)
	for i, h := range rows[headerIdx] {
		actualHeaders[strings.ToLower(strings.TrimSpace(h))] = i
	}

	cols := cfg.Columnas
	columnFields := map[string]string{
		"Tipo Transacción":      cols.TipoTransaccion,
		"Cliente":               cols.Cliente,
		"Placa":                 cols.Placa,
		"Valor Actual":          cols.ValorActual,
		"% Interés":             cols.PorcentajeInteres,
		"Valor Interés Mensual": cols.ValorInteresMensual,
		"Vencimiento Interés":   cols.VencimientoInteres,
		"Días Corridos":         cols.DiasCorridos,
		"Valor Interés Vencido": cols.ValorInteresVencido,
		"Saldo Actual":          cols.SaldoActual,
		"Teléfono":              cols.Telefono,
	}

	for label, colName := range columnFields {
		normal := strings.ToLower(strings.TrimSpace(colName))
		if _, ok := actualHeaders[normal]; !ok {
			r.ColumnsMissing = append(r.ColumnsMissing, fmt.Sprintf("%s ('%s')", label, colName))
		}
	}

	if len(r.ColumnsMissing) > 0 {
		r.Warnings = append(r.Warnings, fmt.Sprintf("Columnas no encontradas en el Excel: %s", strings.Join(r.ColumnsMissing, ", ")))
	}

	phoneCol, phoneFound := actualHeaders[strings.ToLower(strings.TrimSpace(cols.Telefono))]
	nameCol, _ := actualHeaders[strings.ToLower(strings.TrimSpace(cols.Cliente))]

	for i := headerIdx + 1; i < len(rows); i++ {
		row := rows[i]
		if len(row) == 0 {
			continue
		}
		r.TotalRows++

		if phoneFound {
			phone := excel.CleanPhone(excel.GetCell(row, phoneCol))
			if phone == "" {
				r.RowsWithoutPhone++
			} else {
				r.RowsWithPhone++
				if r.SamplePhone == "" {
					r.SamplePhone = phone
				}
			}
		}

		if nameCol >= 0 && nameCol < len(row) && strings.TrimSpace(row[nameCol]) == "" && phoneFound {
			phone := excel.CleanPhone(excel.GetCell(row, phoneCol))
			if phone != "" {
				r.Warnings = append(r.Warnings, fmt.Sprintf("Fila %d: tiene teléfono pero el nombre está vacío", i+1))
			}
		}
	}

	if r.TotalRows == 0 {
		r.Warnings = append(r.Warnings, "No hay filas de datos después de la fila de encabezado")
	}

	if phoneFound && r.RowsWithPhone == 0 {
		r.Warnings = append(r.Warnings, "Ninguna fila tiene un número de teléfono válido")
	}

	if r.RowsWithoutPhone > 0 {
		r.Warnings = append(r.Warnings, fmt.Sprintf("%d filas sin teléfono válido (serán omitidas al notificar)", r.RowsWithoutPhone))
	}

	if !phoneFound {
		r.Errors = append(r.Errors, "Columna de teléfono no encontrada — no se podrá notificar a nadie")
		r.Valid = false
	}

	return r
}

func (r *Result) String() string {
	var b strings.Builder
	b.WriteString("VALIDACIÓN DE CONFIGURACIÓN\n")

	if len(r.Errors) > 0 {
		b.WriteString("ERRORES:\n")
		for _, e := range r.Errors {
			b.WriteString(fmt.Sprintf("  • %s\n", e))
		}
		b.WriteString("\n")
	}

	if len(r.Warnings) > 0 {
		b.WriteString("ADVERTENCIAS:\n")
		for _, w := range r.Warnings {
			b.WriteString(fmt.Sprintf("  • %s\n", w))
		}
		b.WriteString("\n")
	}

	b.WriteString("RESUMEN:\n")
	b.WriteString(fmt.Sprintf("  Archivo:        %s\n", map[bool]string{true: "Encontrado", false: "No encontrado"}[r.FileExists]))
	b.WriteString(fmt.Sprintf("  Columnas OK:    %d/11\n", 11-len(r.ColumnsMissing)))
	b.WriteString(fmt.Sprintf("  Filas de datos: %d\n", r.TotalRows))
	b.WriteString(fmt.Sprintf("  Con teléfono:   %d\n", r.RowsWithPhone))
	b.WriteString(fmt.Sprintf("  Sin teléfono:   %d\n", r.RowsWithoutPhone))
	if r.SamplePhone != "" {
		b.WriteString(fmt.Sprintf("  Tel. ejemplo:   %s\n", r.SamplePhone))
	}
	b.WriteString("\n")

	if r.Valid {
		b.WriteString("TODO CORRECTO — La configuración es válida\n")
	} else {
		b.WriteString("HAY ERRORES — Corrige los problemas antes de ejecutar\n")
	}

	return b.String()
}

File: notiflow-settings-ui/src/validator.rs
use std::collections::HashMap;
use std::fs;

use calamine::{Data, Reader, Xlsx, open_workbook};

use crate::model;

pub struct ValidationResult {
    pub is_valid: bool,
    pub file_exists: bool,
    pub columns_missing: Vec<String>,
    pub total_rows: u32,
    pub rows_with_phone: u32,
    pub rows_without_phone: u32,
    pub sample_phone: String,
    pub errors: Vec<String>,
    pub warnings: Vec<String>,
}

fn cell_string(cell: &Data) -> String {
    match cell {
        Data::String(s) => s.clone(),
        Data::Empty => String::new(),
        other => other.to_string(),
    }
}

pub fn validate(data: &model::Data) -> ValidationResult {
    let mut r = ValidationResult {
        is_valid: true,
        file_exists: false,
        columns_missing: Vec::new(),
        total_rows: 0,
        rows_with_phone: 0,
        rows_without_phone: 0,
        sample_phone: String::new(),
        errors: Vec::new(),
        warnings: Vec::new(),
    };

    let path = &data.excel.path;
    if path.is_empty() {
        r.errors.push("La ruta del archivo Excel está vacía".into());
        r.is_valid = false;
        return r;
    }

    match fs::metadata(path) {
        Ok(m) if m.is_file() => r.file_exists = true,
        _ => {
            r.errors.push(format!("Archivo no encontrado: {}", path));
            r.is_valid = false;
            return r;
        }
    }

    let mut workbook: Xlsx<_> = match open_workbook(path) {
        Ok(wb) => wb,
        Err(e) => {
            r.errors.push(format!("Error abriendo Excel: {}", e));
            r.is_valid = false;
            return r;
        }
    };

    let sheet_names = workbook.sheet_names();

    if sheet_names.is_empty() {
        r.errors.push("El archivo Excel no tiene hojas".into());
        r.is_valid = false;
        return r;
    }

    let range = match workbook.worksheet_range(&sheet_names[0]) {
        Ok(rng) => rng,
        Err(e) => {
            r.errors
                .push(format!("Error leyendo hoja '{}': {}", sheet_names[0], e));
            r.is_valid = false;
            return r;
        }
    };

    let header_row = if data.excel.header_row == 0 {
        1
    } else {
        data.excel.header_row as usize
    };

    let rows: Vec<Vec<Data>> = range.rows().map(|r| r.to_vec()).collect();

    if rows.len() < header_row {
        r.errors.push(format!(
            "El archivo tiene {} filas, se necesita al menos {}",
            rows.len(),
            header_row
        ));
        r.is_valid = false;
        return r;
    }

    let mut actual_headers: HashMap<String, usize> = HashMap::new();
    for (i, cell) in rows[header_row - 1].iter().enumerate() {
        let h = cell_string(cell).trim().to_lowercase();
        if !h.is_empty() {
            actual_headers.insert(h, i);
        }
    }

    let cols: Vec<(&str, &str)> = vec![
        ("Tipo Transacción", &data.columnas.tipo_transaccion),
        ("Cliente", &data.columnas.cliente),
        ("Placa", &data.columnas.placa),
        ("Valor Actual", &data.columnas.valor_actual),
        ("% Interés", &data.columnas.porcentaje_interes),
        (
            "Valor Interés Mensual",
            &data.columnas.valor_interes_mensual,
        ),
        ("Vencimiento Interés", &data.columnas.vencimiento_interes),
        ("Días Corridos", &data.columnas.dias_corridos),
        (
            "Valor Interés Vencido",
            &data.columnas.valor_interes_vencido,
        ),
        ("Saldo Actual", &data.columnas.saldo_actual),
        ("Teléfono", &data.columnas.telefono),
    ];

    for (label, col_name) in &cols {
        let normal = col_name.trim().to_lowercase();
        if !actual_headers.contains_key(&normal) {
            r.columns_missing
                .push(format!("{} ('{}')", label, col_name));
        }
    }

    if !r.columns_missing.is_empty() {
        r.warnings.push(format!(
            "Columnas no encontradas: {}",
            r.columns_missing.join(", ")
        ));
    }

    let phone_col = cols
        .iter()
        .position(|(label, _)| *label == "Teléfono")
        .and_then(|_| {
            let col_name = &data.columnas.telefono;
            let normal = col_name.trim().to_lowercase();
            actual_headers.get(&normal).copied()
        });

    for i in header_row..rows.len() {
        let row = &rows[i];
        if row.is_empty() || row.iter().all(|c| matches!(c, Data::Empty)) {
            continue;
        }
        r.total_rows += 1;

        if let Some(col) = phone_col {
            let phone = clean_phone(&cell_string(row.get(col).unwrap_or(&Data::Empty)));
            if phone.is_empty() {
                r.rows_without_phone += 1;
            } else {
                r.rows_with_phone += 1;
                if r.sample_phone.is_empty() {
                    r.sample_phone = phone;
                }
            }
        }
    }

    if r.total_rows == 0 {
        r.warnings
            .push("No hay filas de datos después del encabezado".into());
    }

    if phone_col.is_some() && r.rows_with_phone == 0 {
        r.warnings
            .push("Ninguna fila tiene un número de teléfono válido".into());
    }

    if r.rows_without_phone > 0 {
        r.warnings.push(format!(
            "{} fila(s) sin teléfono (serán omitidas)",
            r.rows_without_phone
        ));
    }

    if phone_col.is_none() {
        r.errors.push("Columna de teléfono no encontrada".into());
        r.is_valid = false;
    }

    r
}

fn clean_phone(s: &str) -> String {
    s.chars().filter(|c| c.is_ascii_digit()).collect()
}

impl ValidationResult {
    pub fn summary(&self) -> String {
        let mut s = String::new();

        s.push_str("   VALIDACIÓN DE CONFIGURACIÓN\n");

        if !self.errors.is_empty() {
            s.push_str("❌ ERRORES:\n");
            for e in &self.errors {
                s.push_str(&format!("  • {}\n", e));
            }
            s.push('\n');
        }

        if !self.warnings.is_empty() {
            s.push_str("ADVERTENCIAS:\n");
            for w in &self.warnings {
                s.push_str(&format!("  • {}\n", w));
            }
            s.push('\n');
        }

        s.push_str("📊 RESUMEN:\n");
        s.push_str(&format!(
            "  Archivo:        {}\n",
            if self.file_exists {
                "✅ Encontrado"
            } else {
                "❌ No encontrado"
            }
        ));
        s.push_str(&format!(
            "  Columnas OK:    {}/11\n",
            11 - self.columns_missing.len()
        ));
        s.push_str(&format!("  Filas de datos: {}\n", self.total_rows));
        s.push_str(&format!("  Con teléfono:   {}\n", self.rows_with_phone));
        s.push_str(&format!("  Sin teléfono:   {}\n", self.rows_without_phone));
        if !self.sample_phone.is_empty() {
            s.push_str(&format!("  Tel. ejemplo:   {}\n", self.sample_phone));
        }
        s.push('\n');

        if self.is_valid {
            s.push_str("✅ TODO CORRECTO\n");
        } else {
            s.push_str("❌ HAY ERRORES — Corrige antes de ejecutar\n");
        }

        s
    }
}
